man/pmenu.1
.TH PMENU 1 "630 MTG"
.SH NAME
pmenu \- program the 630 MTG Windowproc Button 2 menu
.SH SYNOPSIS
.B  pmenu
.RB [ \-f " file]"
[ \-c ] [ \-C ]
.SH DESCRIPTION
\f2Pmenu\f1
will program the Button 2 windowproc menu for the layers
window it is invoked in.
\f2Pmenu\f1 reads a configuration file (\f2.mtgmenurc\f1 or the file
specified by the \f2\-f\f1 flag) that describes the menus to be added.
Each line of the configuration file consists of three arguements:
.ftCW
.ti +5
.br
depth "item" "string"
.br
.ft1
\f2Depth\f1 can have values from 0 (the top depth menu) to 3 (three levels down) and
describes where the item is to be placed.
Quotes must enclose both the \f2item\f1 and \f2string\f1 fields.
The \f2"item"\f1 field
will be printed as an entry in the menu at that
depth. The \f2"string"\f1 field
are the characters that will be transmitted to the
host when the corresponding \f2"item"\f1 is selected with the mouse and
Button 2.
A \f2\\n\f1 placed in the \f2"string"\f1 field
allows a newline to be transmitted to the host.
When Button 2 is used to select an item either a string is transmitted
to the host or a sub-menu is displayed.
.PP
\f2Pmenu\f1 creates menu entries in the order they appear in the configuration
file. When a line in the configuration file
at depth n is followed by one at depth n+1, the depth n
item becomes a pointer to the a sub-menu with the depth n+1 item as the top
entry. Any \f2"string"\f1 bound to the depth n item will be ignored.
When a line in the configuration file at depth n is followed by one
at depth n-1, the depth n-1 line is placed at the end of the menu that
contains the pointer to the depth n sub-menu.
For example the following is a valid configuration file:
.nf
.ftCW
.in +5
.br
0 "vi"
1 "up" "k"
1 "down" "j"
1 "left" "h"
1 "right" "l"
0 "rlogin"
1 "ucsd"
2 "sdcc7" "rlogin sdcc7\\n"
2 "sdcc10" "rlogin sdcc10\\n"
1 "berkeley"
2 "monet" "rlogin monet.berkeley.edu\\n"
2 "okeeffe" "rlogin okeeffe.berkeley.edu\\n"
2 "ucbvax" "rlogin ucbvax.berkeley.edu\\n"
.br
.ft1
.fi
.in -5
.PP
This creates a pointer \f2vi\f1 to a sub-menu with four entries in it,
and a pointer \f2rlogin\f1 to a sub-menu with pointers to two sub-menus.
.PP
The command-line options are:
.TP 8
.BI \-f " file"
Specifies that "file" be used as an initial configuration file
instead of the default "$HOME/.mtgmenurc" file.
.TP 8
.B \-c
Specifies that all menus added to Button 2 in this window be cleared before
reading the configuration file.
.TP 8
.B \-C
Specifies that all menus added to Button 2 in this window be cleared (the
configuration is not read).
.PP
See the 630 Multi-Tasking Graphical Terminal User's Guide for more information
about programming the Windowproc Button 2 menu.


HISTORY
This file contains brief descriptions of changes made to the pseudo-tty
layers in chronological order.  This file was created 4/19/93 so dates
before that time are approximate.  Unless otherwise noted, all changes were
made by Dave Dykstra.

1991
====
10/1	Keith Muller's version of layers retrieved from anonymous ftp ucsd.edu 
	at pub/dmdlayers.tar.Z.

12/1	Modified openchan() libwindows call to return a pseudo-tty instead of
	a socket; this supports starting a program in a window from another
	process such as "agent !".  Also removed openrwchan() as it is no
	longer necessary.

1992
====
2/1	Added calls to signal() at the end of signal handling routines to
	reset signal handlers in case layers is compiled with /usr/5bin/cc
	on a sun.

4/23	Strip off eigth bit of TIOCUCNTL calls because the R&D Unix dmdld
	had turned the uppermost bit on for some reason.

5/1	Addition of one second delay after switching to layers mode to
	allow time for the terminal to initialize.

7/1	Allowed layers to come up even if not setuid-root.  When not setuid-
	root, can't write /etc/utmp unless it is world writable, and can't
	change ownership of pseudo-ttys.

9/1	Strip off 8th bit when inquiring the terminal for it's escape
	sequence identifier.  This allows getting into layers no matter
	what the parity setting.  Also allowed for missing square bracket
	in the returned escape sequence in case the wproc terminal controls
	option is set to 8 bits.  Will print out the id strings in a
	readable format if they are not recognized.

10/1	No longer requires empty files in $DMD/lib/layersys for each terminal
	id string to get into layers.

12/1	Added NO_TIOCUCNTL to support machine types on which the TIOCUCNTL
	ioctl on pseudo-ttys is not supported.

12/9	Added getagentanswer() libwindows call to get access to the return
	buffer from an agent call.  This is used by the agent(1) program to
	display the return buffer in ascii for agent calls like A_ROMVERSION.

12/14	Added '-V' option to print out version.
	Made lyrscfg.h to contain defines for different operating system
	configurations.
	Added support for mips-based OS's, specifically StarServer-FT,
	thanks to many changes by John Ampe.  This OS is a combination of
	System V and BSD.
	One notable addition is USE_NAMED_PIPE to use named pipes to
	communicate with support programs rather than unix-domain sockets.
	Another is support for system V utmp file format, although it is
	known to be buggy.
	Changed default shell to /bin/sh instead of /bin/csh because csh is
	not available on all systems.
	Added support for AT&T SVR4, USE_PTMX.

1993
====
2/1	Added -e option to force on hex encoding mode the same way that
	the environment variable DMDLOAD=hex does.

3/1	Increased number of pseudo-ttys supported from 96 to 256 on
	machines using the original bsd naming conventions
	(/dev/[pt]ty[pqrstuvwxyzPQRSTU][0-f]).

	Change to pass unrecognized agent calls on to the terminal rather
	than rejecting them immediately.

3/10	Added more portability defines for HP-PA thanks to Mark Larsen.

4/19	Added beginning of support for Solaris from Bob Pietkivitch.
	Added John Lash's '-m' option for larger packet size support on 730s.
	Got to compile without warnings on NCR 3000 SVR4 with both the AT&T
	and MetaWare compilers.
	Removed old sunos pty driver patches from misc directory.

4/21	Change to always return the result from an agent call in the agent
	answer buffer.
	Add flow control on agent calls to make sure all data from the
	pseudo-tty is drained before processing the agent call.
	Add support for A_JBOOT, A_JTERM, and A_JZOMBOOT even if on a system
	that has TIOCUCNTL.  The purpose is to be able to use the above
	flow control and to provide for transition to operating system
	versions that don't support TIOCUCNTL (such as sun -> solaris).
	Change libwindows to send the $DMDCHAN with the request if set
	before looking up ttyname() because it's much faster.

4/22	Add matchtty() call to match pty name on major/minor number if a
	match is not found by name.  This is for systems that have two /dev
	names for the same ptys: the one that layers sees may not be the
	same one that ttyname() sees, and having layers() call ttyname() is
	much slower.
	Change to print error message to opened window if can't allocate
	pty; previously the window was silently deleted.
	Made a more reasonable algorithm for determining timeout times
	based on the baud rate.  If the user does not specify '-b' baud
	rate, the maximum is assumed to be 4800 baud to allow for network
	delays and other delays; the timeout for 4800 baud is now 6 seconds
	for transmit and 4 seconds for receive, double the preceding minimum
	(and the mimimum was usually selected).
	Eliminated PXTIMEOUT and RXTIMEOUT defines and instead use the value
	for 1200 baud in the table at initialization time.
	Put in a one second delay when forking a process to create a new
	window to allow time for the process to open the pty before layers
	tries to read it.

4/27	Ignore leading blanks in a '-f' file.
	Fixed a problem with starting up multiple dmdld's from a '-f' file
	(only for libwindows programs that use A_JBOOT, not those using the
	JBOOT TIOCUCNTL ioctl): too many packets get sent on channel 0
	before being acked so some of them fail allocating a buffer to send
	the JBOOT message.  The fix is to delay agent messages if there are
	unacked messages on channel 0, using the agentblock flag.

4/28	Got rid of wtinit.  It was just a simple program that sent the
	startup string for layers, with hex encoding on or off.

5/4	Removed setenv.c and getenv.c.  Just use standard putenv/getenv instead.

5/5     Created 'grantpt' and 'movelogin' library programs.  These are small
	programs that are to be installed setuid root to grant ownership
	of pseudo ttys and to move entries around in the utmp file, respect-
	ively.  Now layers itself no longer needs to be setuid root.
	'Grantpt' is not needed on SVR4 because a similar program comes

misc/pmenu.c
/*
 * pmenu: a utility for programming menus feature of windowproc on a 630 MTG
 *        each menu is bound to the window it is run in.
 * usage: pmenu [-c|-C] [-f menufile]
 *
 *	  menufile looks like:
 *		depth "item" "string"
 *	  note that the " around item and string are required. string is
 *	  optional for upper depth links.
 *	  a '\n' in the "string" will be replaced by a newline character
 */

#include <stdio.h>
#include "windowproc.h"
#include "lyrscfg.h"
#define USEAGE	"Usage: %s [-c|-C][-f menufile]\n"

#ifdef NO_INDEX
#define index strchr
#endif

main(argc, argv)
	int argc;
	char **argv;
{
	register char *ptr;
	register char *iptr;
	register char *item;
	register char *string;
	register int retval;
	int fflag = 0;
	int cflag = 0;
	int quit = 0;
	FILE *rcfile;
	char inbuf[BUFSIZ];
	char *mtgmenurc = inbuf;
	char *home;
	int count;
	int depth;
	extern char *optarg;
	extern int optind;
	extern char *getenv();
	extern char *index();
	
	if (((home = getenv("HOME")) == (char *) 0) || (*home == '\0'))
		mtgmenurc = MTGMENURC;
	else
		sprintf(inbuf, "%s/%s", home, MTGMENURC);
		
	while ((retval = getopt(argc, argv, "Ccf:")) != EOF) {
		switch(retval) {
		case 'f':
			mtgmenurc = optarg;
			fflag++;
			break;
		case 'C':
			quit++;
			/*
			 * fall through
			 */
		case 'c':
			/*
			 * clear all menus
			 */
			cflag++;
			break;
		case '?':
		default:
			fprintf(stderr, USEAGE, argv[0]);
			exit(1);
		}
	}

	if (cflag)
		printf(PRGRMENU, 0, 0, 0, "", "");
	if (quit)
		exit(0);
	if ((rcfile = fopen(mtgmenurc, "r")) == NULL) {
		if (fflag == 0)
			exit(0);
		fprintf(stderr, "%s: unable to open %s\n",argv[0],mtgmenurc);
		exit(1);
	}
	while (fgets(inbuf, BUFSIZ, rcfile) != NULL) {
		if ((ptr = index(inbuf, '"')) == (char *)0)
			continue;
		*ptr++ = '\0';
		if (((depth = atoi(inbuf)) < 0) || (depth > 3))
			continue;
		item = ptr;
		if ((ptr = index(item, '"')) == (char *)0)
			continue;
		*ptr++ = '\0';
		if ((string = index(ptr, '"')) == (char *)0)
			string = "";
		else
			string++;

		if ((ptr = index(string, '"')) != (char *)0)
			*ptr = '\0';
		for (ptr = string; *ptr != '\0'; ptr ++) {
			if (*ptr != '\\')
				continue;
			if (*(ptr+1) != 'n')
				continue;
			*ptr++ = '\n';
			for (iptr = ptr; *iptr != '\0'; iptr++)
				*iptr = *(iptr+1);
		}
		printf(PRGRMENU,strlen(item),strlen(string),depth,item,string);
	}
	exit(0);
}
